### ISSUE TITLE
[🐛 Bug]: dynamic page with catch all route overrides all API routes

### ISSUE THREAD

**Issue by fong** [cloudflare/next-on-pages#593]
### next-on-pages environment related information

```
System:
        Platform: darwin
        Arch: arm64
        Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:12 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T8103
        CPU: (8) arm64 Apple M1
        Memory: 16 GB
        Shell: /bin/zsh
Package Manager Used: npm (10.2.3)

Relevant Packages:
        @cloudflare/next-on-pages: 1.8.2
        vercel: N/A
        next: 14.0.4
```

### Description

When using a catch all page routes on the top level (e.g. `/pages/[...slug].tsx` or  `/app/[...slug]/page.tsx`), all API routes incorrectly route to the catch all page instead of hitting the API endpoint.

Have also tried local dev server/builds and deployment to Vercel with the same codebase, but haven't been able to reproduce it in those environments, so it looks like it only affects `@cloudflare/next-on-pages` builds (both local `npx @cloudflare/next-on-pages` and on Cloudflare Pages deployment).

Problem occurs in both `app` and `page` routers.

### Reproduction

Regular catch all page (using `pages` router)
https://dynamic-routes-api-404.pages.dev/home/page/
https://dynamic-routes-api-404.vercel.app/home/page/

API route (using `pages` router):
https://dynamic-routes-api-404.pages.dev/api/hello/ (loads catch all page)
https://dynamic-routes-api-404.vercel.app/api/hello/ (loads API route)

Regular catch all page (using `app` router)
https://dynamic-routes-api-404-app.pages.dev/home/page/
https://dynamic-routes-api-404-app.vercel.app/home/page/

API route (using `app` router):
https://dynamic-routes-api-404-app.pages.dev/api/hello/  (loads catch all page)
https://dynamic-routes-api-404-app.vercel.app/api/hello/ (loads API route)

Repos for repro: 
Pages: https://github.com/fong/dynamic-routes-api-404 
App: https://github.com/fong/dynamic-routes-api-404-app

### Pages Deployment Method

Pages CI (GitHub/GitLab integration)

### Pages Deployment ID

b61f75ac

### Would you like to help?

- [ ] Would you like to help fixing this bug?

**Comment by dario-piotrowicz** [comment]
Thanks for the issue @fong 🙂 

And thanks so so much for the awesome reproductions!!! 💪 🚀 

I've implemented a fix for the issue (#594), if you want please feel free to try out the [PR's prerelease](https://github.com/cloudflare/next-on-pages/pull/594#issuecomment-1855642666) and let me know if you're happy with the fix 🙂

**Comment by fong** [comment]
Gave it a try on the affected project and the fix works great!

Cheers for the quick turnaround!
